### ISSUE TITLE
`yargs.exit` takes an error as an argument but doesn't seem to output said error

### ISSUE THREAD

**Issue by jameswomack** [yargs/yargs#1725]
I ❤️  Yargs, thank you to everyone who's worked on it to make it how great it is today!

I'm wanting to implement a canonical way for our CLI commands to exit out in the case or error. After using a custom solution w/ `process.exit` & a Winston error log transport, a colleague pointed out `yargs.exit`. Given its method signature, I expected `yargs.exit` to output some sort of standardized error message as it exited the process, but that hasn't been the case in my experience. Is there a way to get `yargs.exit` to do so? If not, what's the canonical way to exit w/ an error message in a Yargs-based CLI?

The Yargs source:
```js
self.exit = (code, err) => {
        hasOutput = true;
	exitError = err;
        if (exitProcess)
            process.exit(code);
};
```

My source:
```js
const createFailureMessage = (prefix, errorMessage) =>
	chalk`${prefix} ({italic ${errorMessage}}).`;

const exitWithCode = (exitCode, errorMessagePrefix, errorMessage = exitCode) => {
	const failureMessage = createFailureMessage(errorMessagePrefix, errorMessage);
	return yargs.exit(exitCode, failureMessage);
};

exitWithCode(1, 'Failed to reach John Tejada concert', errorMessage);
```

Expected output:
`Failed to pack project (Cannot read property '0' of undefined).`

Actual output:
` `

**Comment by bcoe** [comment]
@jameswomack hey 👋 sorry it took me a while to triage this. I would love to get to t he bottom of what's biting you here ... I'm  actually surprise that using `yargs.exit()` didn't work as  expected.

Perhaps send a failing test? and I can take a look ASAP?

**Comment by bcoe** [comment]
@jameswomack would an API like this do the trick for you:

```js
const yargs = require('./')()
  yargs.command('cmd', 'description', () => {}, (argv) => {
    if (!argv.blerg) {
      yargs.fail('you need to set blerg')
    } else {
      console.info('all is well')
    }
  })

yargs.parse(process.argv.slice(2))
```

It seems like it would be a pretty reasonable addition to how `fail` works.

**Comment by bcoe** [comment]
This outputs:

```bash
foo.cjs cmd

description

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]

you need to set blerg
```

**Comment by jameswomack** [comment]
@bcoe that'd work. I'll try and cook up a test for `yargs.exit`

**Comment by bcoe** [comment]
@jameswomack I would happily take a patch, I did some experimentation, but I want to make sure it meets your use case before we land the feature.

**Comment by Dunky13** [comment]
I am facing a similar issue, when I run a command as an ES6 module
and the handler is defined as follows:
```ts
handler: async function (argv) {
    if (__dirname.includes('bind')) {
      Yargs.exit(1, new Error('Should not be ran from bin but from lib'))
    }
    console.log('Cool')
  },
```

The output is blank, but it exits with exit code 1.

**Comment by mkilpatrick** [comment]
Any possibility of this being addressed? It's still an issue.

**Comment by shadowspawn** [comment]
Yes, but has not had a lot of likes so not highest priority for this maintainer. 

The high level goal is a way to elegantly exit with a user-visible error string using Yargs machinery.

(I notice the `exit` implementation does store the error which might get used later if `exitProcess` is false. I have not checked what happens in that case, but I suspect not a solution as such.)

**Comment by shadowspawn** [comment]
This issue has not had any activity in over six months. It isn't likely to get acted on due to this report.

Feel free to open a new issue if it comes up again, with new information and renewed interest.
 
Thank you for your contributions.
